lib/active_stash/index_dsl.rb

```diff
@@ -0,0 +1,285 @@
+module ActiveStash
+  # Implements the DSL for adding searchable encrypted indexes to ActiveRecord
+  # models that include ActiveStash::Search.
+  #
+  #   stash_index do
+  #     auto :email, :first_name, :last_name
+  #     unique :email
+  #     range :created_at, :updated_at
+  #     exact :gender
+  #
+  #     match :description, filter_term_bits: 512
+  #
+  #     match_all :first_name, :last_name, :email
+  #
+  #     index_assoc :patient do
+  #       auto :email, :first_name, :last_name
+  #     end
+  #   end
+  #
+  class IndexDSL
+    def initialize(model_class, path = [], reflector = nil)
+      @model_class = model_class
+      @reflector = reflector || Reflector.new(@model_class)
+      @path = path || []
+      @is_in_association = path.size > 0
+      @indexes = []
+      @associations = []
+      @unique_fields = []
+      @callback_registration_handlers = []
+    end
+
+    # Performs the following checks and operations before returning the final
+    # indexes.
+    #
+    # 1. Validates that there are not multiple index definitions of the same
+    # type for the same field.
+    #
+    # 2. Validates that there are not multiple associated indexes defined for
+    # the same association.
+    #
+    # 3. Processes `unique` indexes by either:
+    #
+    #    - Updating existing `range` and `exact` indexes for a field to enforce
+    #    uniqueness
+    #
+    #    - Creating an `exact` or `range` index for the field if one does not
+    #    already exist (`string` and `text` fields generate `exact` indexes,
+    #    everything else generates a `range` index).
+    #
+    def finalize!
+      return @finalized_config if @finalized_config
+      validate_no_fields_with_duplicate_index_definitions!
```

Comment:
It may well make sense for a field to have multiple `match` indexes on a single field. They could be useful in different cases. Maybe we make that a future enhancement.

Reply:
That's a good point, but we currently don't provide a way (in ActiveStash) for a user to name indexes in order for them to direct a query expression to the correct index. All the name mangling is automatic.